### ISSUE TITLE
`solve_cubic` returns 1.0 for every linear equation and zeros for `bx² + d`; `solve_quadratic` returns NaN for `a = 0`

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4873]
`solve_cubic` handles a zero leading coefficient with two masks. For `a = b = 0, c ≠ 0` (a linear equation), it writes the constant `torch.tensor(1.0)` into slot 0, whatever `c` and `d` are. For `a = 0, b ≠ 0, c ≠ 0`, it forwards the quadratic to `solve_quadratic`. `a = 0, b ≠ 0, c = 0` (`bx² + d`) matches neither mask and returns zeros. `solve_quadratic` divides by `2a` without a guard and returns `[nan, -inf]` for `2x − 6`. `solve_quartic` sends `|a| < tol` to `solve_cubic` and inherits both defects. `np.roots` strips leading zeros and returns `[3]`, `[-5]` and `[2, -2]` for `2x − 6`, `x + 5` and `x² − 4`.

## Repro

```python
import torch
from kornia.geometry.solvers import solve_cubic, solve_quadratic, solve_quartic

cases = [
    (solve_quadratic, [0.0, 2.0, -6.0], "2x - 6, root 3"),
    (solve_cubic, [0.0, 0.0, 2.0, -6.0], "2x - 6, root 3"),
    (solve_cubic, [0.0, 0.0, 1.0, 5.0], "x + 5, root -5"),
    (solve_cubic, [0.0, 0.0, 1.0, -1.0], "x - 1, root 1 (the existing test's case)"),
    (solve_cubic, [0.0, 1.0, 0.0, -4.0], "x^2 - 4, roots 2, -2"),
    (solve_cubic, [0.0, 1.0, -1.0, -6.0], "x^2 - x - 6, roots 3, -2 (handled)"),
    (solve_quartic, [0.0, 0.0, 0.0, 2.0, -6.0], "2x - 6, root 3"),
    (solve_quartic, [0.0, 0.0, 2.0, 0.0, -8.0], "2x^2 - 8, roots 2, -2"),
]
for fn, c, what in cases:
    print(f"{fn.__name__}({c}) -> {fn(torch.tensor([c])).tolist()[0]}   # {what}")
```

```text
solve_quadratic([0.0, 2.0, -6.0]) -> [nan, -inf]   # 2x - 6, root 3
solve_cubic([0.0, 0.0, 2.0, -6.0]) -> [1.0, 0.0, 0.0]   # 2x - 6, root 3
solve_cubic([0.0, 0.0, 1.0, 5.0]) -> [1.0, 0.0, 0.0]   # x + 5, root -5
solve_cubic([0.0, 0.0, 1.0, -1.0]) -> [1.0, 0.0, 0.0]   # x - 1, root 1 (the existing test's case)
solve_cubic([0.0, 1.0, 0.0, -4.0]) -> [0.0, 0.0, 0.0]   # x^2 - 4, roots 2, -2
solve_cubic([0.0, 1.0, -1.0, -6.0]) -> [3.0, -2.0, 0.0]   # x^2 - x - 6, roots 3, -2 (handled)
solve_quartic([0.0, 0.0, 0.0, 2.0, -6.0]) -> [1.0, 0.0, 0.0, 0.0]   # 2x - 6, root 3
solve_quartic([0.0, 0.0, 2.0, 0.0, -8.0]) -> [0.0, 0.0, 0.0, 0.0]   # 2x^2 - 8, roots 2, -2
```

kornia `main` 7ddf731bd, torch 2.14.0 (and 2.5.1: same output), CPU float32, macOS arm64.

The existing test cannot detect the defect: `TestCubicSolver::test_solve_quadratic_in_cubic` covers the linear branch with `[0, 0, 1, -1]`, i.e. `x − 1`, whose root is exactly the hard-coded `1.0`. The output is the same on MPS float32. Missing real roots are padded with `0.0` by design (#4862 tracks the consequence of that padding in `run_7point`).

## Classification

Clearly broken output: wrong finite roots (`1.0` for `2x − 6`) and non-finite ones (`nan`, `-inf`) with no error.

Idiom 4 (degenerate or singular input silently producing garbage instead of raising), and the linear branch writes a wrong constant.

## Disposition

Focused fix, welcome as a PR: the linear branch returns `-d / c`; the `bx² + d` case (and every `a = 0, b ≠ 0` row) goes to `solve_quadratic`; `solve_quadratic` solves `a = 0, b ≠ 0` as linear, padding the second slot per the existing convention. Change the linear test case to `2x − 6` (root 3), which the hard-coded `1.0` cannot satisfy.

Found by the batch-8 conventions audit (two-view geometry); a `test_wart_*` pin in the docs PR flips when the fix lands.

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).
